File: fc-chat/demo/part3/ch3/02-6-message-system-add-connection/src/main/java/net/prostars/messagesystem/handler/websocket/RequestDispatcher.java
package net.prostars.messagesystem.handler.websocket;

import jakarta.annotation.PostConstruct;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.Map;
import net.prostars.messagesystem.dto.websocket.inbound.BaseRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.ListableBeanFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.socket.WebSocketSession;

@Component
@SuppressWarnings({"rawtypes", "unchecked"})
public class RequestDispatcher {

  private static final Logger log = LoggerFactory.getLogger(RequestDispatcher.class);

  private final Map<Class<? extends BaseRequest>, BaseRequestHandler<? extends BaseRequest>>
      handlerMap = new HashMap<>();
  private final ListableBeanFactory listableBeanFactory;

  public RequestDispatcher(ListableBeanFactory listableBeanFactory) {
    this.listableBeanFactory = listableBeanFactory;
  }

  public <T extends BaseRequest> void dispatchRequest(
      WebSocketSession webSocketSession, T request) {
    BaseRequestHandler<T> handler = (BaseRequestHandler<T>) handlerMap.get(request.getClass());
    if (handler != null) {
      handler.handleRequest(webSocketSession, request);
      return;
    }
    log.error("Handler not found for request type: {}", request.getClass().getSimpleName());
  }

  @PostConstruct
  private void prepareRequestHandlerMapping() {
    Map<String, BaseRequestHandler> beanHandlers =
        listableBeanFactory.getBeansOfType(BaseRequestHandler.class);
    for (BaseRequestHandler handler : beanHandlers.values()) {
      Class<? extends BaseRequest> requestClass = extractRequestClass(handler);
      if (requestClass != null) {
        handlerMap.put(requestClass, handler);
      }
    }
  }

  private Class<? extends BaseRequest> extractRequestClass(BaseRequestHandler handler) {
    for (Type type : handler.getClass().getGenericInterfaces()) {
      if (type instanceof ParameterizedType parameterizedType
          && parameterizedType.getRawType().equals(BaseRequestHandler.class)) {
        return (Class<? extends BaseRequest>) parameterizedType.getActualTypeArguments()[0];
      }
    }
    return null;
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/handler/websocket/RequestDispatcher.kt
package com.yscorp.withpush.messagesystem.handler.websocket

import com.yscorp.withpush.messagesystem.dto.websocket.inbound.BaseRequest
import jakarta.annotation.PostConstruct
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.ListableBeanFactory
import org.springframework.stereotype.Component
import org.springframework.web.socket.WebSocketSession
import java.lang.reflect.ParameterizedType

@Component
class RequestDispatcher(
    private val listableBeanFactory: ListableBeanFactory
) {

    private val handlerMap: MutableMap<Class<out BaseRequest>, BaseRequestHandler<out BaseRequest>> = mutableMapOf()

    fun <T : BaseRequest> dispatchRequest(webSocketSession: WebSocketSession,
        request: T
    ) {
        val handler = handlerMap[request::class.java] as? BaseRequestHandler<T>
        if (handler != null) {
            handler.handleRequest(webSocketSession, request)
        } else {
            log.error("Handler not found for request type: {}", request::class.simpleName)
        }
    }

    @PostConstruct
    private fun prepareRequestHandlerMapping() {
        val beanHandlers = listableBeanFactory.getBeansOfType(BaseRequestHandler::class.java)
        for (handler in beanHandlers.values) {
            val requestClass = extractRequestClass(handler)
            if (requestClass != null) {
                @Suppress("UNCHECKED_CAST")
                handlerMap[requestClass as Class<out BaseRequest>] = handler as BaseRequestHandler<out BaseRequest>
            }
        }
    }

    private fun extractRequestClass(handler: BaseRequestHandler<*>): Class<out BaseRequest>? {
        for (type in handler.javaClass.genericInterfaces) {
            if (type is ParameterizedType && type.rawType == BaseRequestHandler::class.java) {
                @Suppress("UNCHECKED_CAST")
                return type.actualTypeArguments[0] as? Class<out BaseRequest>
            }
        }
        return null
    }

    companion object {
        private val log: Logger = LoggerFactory.getLogger(RequestDispatcher::class.java)
    }
}

